File: node-vm/cpu.js
const createMemory = require("./memory");
const instructions = require("./instructions");

class CPU {
    constructor(memory) {
        this.memory = memory;

        this.registerNames = [
            "ip", "acc",
            "r1", "r2", "r3", "r4",
            "r5", "r6", "r7", "r8",
        ];

        // 16 bit registers, 2 bytes
        this.registers = createMemory(this.registerNames.length * 2);

        this.registerMap = this.registerNames.reduce((map, name, i) => {
            map[name] = i * 2;
            return map;
        }, {});
    }

    debug() {
        this.registerNames.forEach(name => {
            console.log(`${name}: 0x${this.getRegister(name).toString(16).padStart(4, "0")}`)
        });
        console.log();
    }

    viewMemoryAt(address) {
        const nextEightBytes = Array.from({length: 8}, (_, i) => 
            this.memory.getUint8(address + i)
        ).map(v => `0x${v.toString(16).padStart(2, "0")}`);

        console.log(`${address.toString(16).padStart(4, "0")}: ${nextEightBytes.join(" ")}`);
    }

    getRegister(name) {
        if (!(name in this.registerMap)) {
            throw new Error(`getRegister: no such register '${name}'`);
        }
        return this.registers.getUint16(this.registerMap[name]);
    }

    setRegister(name, value) {
        if (!(name in this.registerMap)) {
            throw new Error(`setRegister: no such register '${name}'`);
        }
        return this.registers.setUint16(this.registerMap[name], value);
    }

    fetch() {
        const nextInstructonAddress = this.getRegister("ip");
        const instruction = this.memory.getUint8(nextInstructonAddress);
        this.setRegister('ip', nextInstructonAddress + 1);
        return instruction;
    }

    fetch16() {
        const nextInstructonAddress = this.getRegister("ip");
        const instruction = this.memory.getUint16(nextInstructonAddress);
        this.setRegister('ip', nextInstructonAddress + 2);
        return instruction;
    }

    execute(instruction) {
        switch (instruction) {
            case instructions.MOV_LIT_REG: {
                const literal = this.fetch16();
                const register = (this.fetch() % this.registerNames.length) * 2;
                this.registers.setUint16(register, literal);
                return;
            }

            case instructions.MOV_REG_REG: {
                const registerFrom = (this.fetch() % this.registerNames.length) * 2;
                const registerTo = (this.fetch() % this.registerNames.length) * 2;
                const value = this.registers.getUint16(registerFrom);
                this.registers.setUint16(registerTo, value);
                return;
            }

            case instructions.MOV_REG_MEM: {
                const registerFrom = (this.fetch() % this.registerNames.length) * 2;
                const address = this.fetch16();
                const value = this.registers.getUint16(registerFrom);
                this.memory.setUint16(address, value);
                return;
            }

            case instructions.MOV_MEM_REG: {
                const address = this.fetch16();
                const registerTo = (this.fetch() % this.registerNames.length) * 2;
                const value = this.memory.getUint16(address);
                this.registers.setUint16(registerTo, value);
                return;
            }

            case instructions.ADD_REG_REG: {
                const r1 = this.fetch();
                const r2 = this.fetch();
                const registerValue1 = this.registers.getUint16(r1 * 2);
                const registerValue2 = this.registers.getUint16(r2 * 2);
                this.setRegister("acc", registerValue1 + registerValue2);
                return;
            }

            case instructions.JMP_NOT_EQ: {
                const literal = this.fetch16();
                const address = this.fetch16();

                if (literal !== this.getRegister("acc")) {
                    this.setRegister("ip", address);
                }

                return;
            }
        }
    }

    step() {
        const instruction = this.fetch();
        return this.execute(instruction);
    }
}

module.exports = CPU;

File: rust-vm/src/cpu.rs
use std::collections::HashMap;

use crate::memory::Memory;
use crate::instructions;

#[derive (Debug)]
pub struct CPU {
    memory: Memory,
    register_names: Vec<String>,
    registers: Memory,
    register_map: HashMap<String, usize>,
}

impl CPU {
    pub fn new(memory: Memory) -> CPU {
        let register_names =  vec![
            "ip".to_string(), "acc".to_string(),
            "r1".to_string(), "r2".to_string(), "r3".to_string(), "r4".to_string(),
            "r5".to_string(), "r6".to_string(), "r7".to_string(), "r8".to_string(),
        ];
        let registers = Memory::new(register_names.len() * 2);
        let register_map =  register_names.iter().enumerate().fold(HashMap::new(), |mut map: HashMap<String, usize>, (i, name)| {
            map.insert(
                name.to_string(), i * 2,
            );
            map
        });

        CPU {
            memory,
            register_names,
            register_map,
            registers,
            
        }
    }

    pub fn debug(&self) {
        for name in &self.register_names {
            println!("{}: {:#06x}", name, self.get_register(name.to_string()));
        }
        println!("");
    }

    pub fn view_memory_at(&self, address: u16) {
        let next_eight_bytes: Vec<String> = vec!["".to_string(); 8].into_iter().enumerate()
            .map(|(i, _v)| {
                format!("{:#04x}", self.memory.get_uint8(address as usize + i))
        }).collect();

        println!("{:#06x}: {}", address, next_eight_bytes.join(" "));
    }

    pub fn get_register(&self, name: String) -> u16{
        // TODO: check for register in registermap
        self.registers.get_uint16(self.register_map[&name])
    }

    fn set_register(&mut self, name: String, value: u16) {
        // TODO: check for register in registermap
        self.registers.set_uint16(self.register_map[&name], value);
    }

    fn fetch(&mut self) -> u8{
        let next_instruction_address = self.get_register("ip".to_string());
        let instruction = self.memory.get_uint8(next_instruction_address.into());
        self.set_register("ip".to_string(), next_instruction_address + 1);
        return instruction;
    }

    fn fetch16(&mut self) -> u16{
        let next_instruction_address = self.get_register("ip".to_string());
        let instruction = self.memory.get_uint16(next_instruction_address.into());
        self.set_register("ip".to_string(), next_instruction_address + 2);
        return instruction;
    }

    fn execute(&mut self, instruction: u8) {
        match instruction {
            instructions::MOV_LIT_REG => {
                let literal = self.fetch16();
                let register = (self.fetch() as usize % self.register_names.len()) * 2;
                self.registers.set_uint16(register, literal);
            },

            instructions::MOV_REG_REG => {
                let register_from = (self.fetch() as usize % self.register_names.len()) * 2;
                let register_to = (self.fetch() as usize % self.register_names.len()) * 2;
                let value = self.registers.get_uint16(register_from.into());
                self.registers.set_uint16(register_to, value);
            },

            instructions::MOV_REG_MEM => {
                let register_from = (self.fetch() as usize % self.register_names.len()) * 2;
                let address = self.fetch16();
                let value = self.registers.get_uint16((register_from).into());
                self.memory.set_uint16(address.into(), value);
            },

            instructions::MOV_MEM_REG => {
                let address = self.fetch16();
                let register_to = (self.fetch() as usize % self.register_names.len()) * 2;
                let value = self.memory.get_uint16(address.into());
                self.registers.set_uint16(register_to, value);
            },

            instructions::ADD_REG_REG => {
                let r1 = self.fetch();
                let r2 = self.fetch();
                let register_value1 = self.registers.get_uint16((r1 * 2).into());
                let register_value2 = self.registers.get_uint16((r2 * 2).into());
                self.set_register("acc".to_string(), register_value1 + register_value2);
            },

            instructions::JMP_NOT_EQ => {
                let literal = self.fetch16();
                let address = self.fetch16();

                if literal != self.get_register("acc".to_string()) {
                    self.set_register("ip".to_string(), address);
                }
            },

            _ => {},
        }
    }

    pub fn step(&mut self) {
        let instruction = self.fetch();
        self.execute(instruction);
    }

}